include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,835 @@
+    static inline constexpr uint32_t COMMAND_SEGMENTS_PER_BLOCK = 256u;
+    static inline constexpr uint32_t MIN_POOL_ALLOC_SIZE = COMMAND_SEGMENT_SIZE;
+
+public:
+    class CCommandSegment;
+
+    class alignas(COMMAND_ALIGNMENT) ICommand
+    {
+    public:
+        virtual ~ICommand() {}
+
+        // static void* operator new(std::size_t size) = delete;
+        static void* operator new[](std::size_t size) = delete;
+        // static void* operator new(std::size_t size, std::align_val_t al) = delete;
+        static void* operator new[](std::size_t size, std::align_val_t al) = delete;
+
+        // static void operator delete  (void* ptr) = delete;
+        static void operator delete[](void* ptr) = delete;
+        static void operator delete  (void* ptr, std::align_val_t al) = delete;
+        static void operator delete[](void* ptr, std::align_val_t al) = delete;
+        static void operator delete  (void* ptr, std::size_t sz) = delete;
+        static void operator delete[](void* ptr, std::size_t sz) = delete;
+        static void operator delete  (void* ptr, std::size_t sz, std::align_val_t al) = delete;
+        static void operator delete[](void* ptr, std::size_t sz, std::align_val_t al) = delete;
+
+        inline uint32_t getSize() const { return m_size; }
+
+    protected:
+        ICommand(uint32_t size) : m_size(size) {}
+
+    private:
+        friend CCommandSegment;
+
+        const uint32_t m_size;
+    };
+
+    template <class CRTP>
+    class NBL_FORCE_EBO IFixedSizeCommand : public IGPUCommandPool::ICommand
+    {
     public:
-        enum E_CREATE_FLAGS : uint32_t
+        template <typename... Args>
+        static uint32_t calc_size(const Args&...)
+        {
+            return sizeof(CRTP);
+        }
+
+    protected:
+        IFixedSizeCommand() : ICommand(calc_size()) {}
+    };
+
+    class alignas(COMMAND_SEGMENT_ALIGNMENT) CCommandSegment
+    {
+    private:
+        struct header_t
         {
-            ECF_NONE = 0x00,
-            ECF_TRANSIENT_BIT = 0x01,
-            ECF_RESET_COMMAND_BUFFER_BIT = 0x02,
-            ECF_PROTECTED_BIT = 0x04
+            core::LinearAddressAllocator<uint32_t> m_commandAllocator;
+            CCommandSegment* m_next;
+        } header;
+
+    public:
+        static inline constexpr uint32_t STORAGE_SIZE = COMMAND_SEGMENT_SIZE - core::roundUp(sizeof(header_t), alignof(ICommand));
+
+        struct Iterator
+        {
+            ICommand* m_cmd = nullptr;
+            // uint32_t m_size = 0;
+            CCommandSegment* m_segment = nullptr;
+
+            inline bool operator!=(const Iterator& other) const { return m_cmd != other.m_cmd; }
+            // TODO(achal): Return Iterator&?
+            // inline void operator++()
+            // {
+            //     m_cmd = reinterpret_cast<ICommand*>(reinterpret_cast<uint8_t*>(m_cmd) + m_size);
+            //     m_size = ;
+            // }
```

Comment:
you don't need `m_size` because the implementation would be
```cpp
inline Iterator& operator++()
{
    if (m_cmd)
    {
       // if `m_cmd` was "one past the end" then the size was wiped to 0 and it won't move
       m_cmd = reinterpret_cast<ICommand*>( reinterpret_cast<uint8_t*>(m_cmd)+m_cmd->getSize());
    }
    return *this;
}
```

Reply:
hmm I wonder what do we even use `m_segment` for?

Reply:
I'm having a hard time understanding what is this check for?
```C++
(m_cmd+1)>=(m_segment+1)
```
More specifically, I cannot understand the following:
1. `CCommandSegment`s are not contiguous in memory so what does `m_segment+1` even mean?
2. When `m_cmd!=old.m_cmd`, it will check the now moved pointer (`m_cmd`)'s next `sizeof(ICommand)` bytes.. but to what end? Like, what is the purpose of it?

Reply:
we scroll to next command in segment, but the last command has the m_size wiped to 0
so then it stops moving
but, if the commands are tightly packed
there might not be a last command that had its m_size wiped to 0 because the location of m_size is past the segments m_data
m_cmd+1 is reinterpret_cast<uint8_t*>(m_cmd)+sizeof(ICommand) so just-after-m_size of m_cmd
m_segment+1 is just after the end of m_data 
so basically (m_cmd+1)>=(m_segment+1) tells you that m_cmd has scrolled so far, that there was no space for its m_size to get wiped to 0

Reply:
Yep you're right.
Resolved.